ext/DDProf/include/ddprof/profiler.hh

```diff
@@ -0,0 +1,87 @@
+#ifndef DDPROF_PROFILER_HH
+#define DDPROF_PROFILER_HH
+
+#include <memory>
+#include <vector>
+
+#include "collector.hh"
+#include "exporter.hh"
+#include "scheduler.hh"
+
+namespace ddprof {
+class profiler {
+    public:
+    enum class status : unsigned {
+        stopped = 0,
+        running = 1,
+    };
+
+    profiler();
+
+    class recorder &get_recorder();
+    void add_collector(std::unique_ptr<class collector> collector);
+    void add_exporter(std::unique_ptr<class exporter> exporter);
+
+    void start() {
+        std::lock_guard<std::mutex> lock{m};
+
+        if (status == status::running) {
+            return;
+        }
+
+        for (auto &collector : collectors) {
+            collector->start();
+        }
+        scheduler.start();
+
+        status = status::running;
+    }
+
+    void stop() {
+        std::lock_guard<std::mutex> lock{m};
+        scheduler.stop();
+
+        for (auto it = collectors.rbegin(); it != collectors.rend(); ++it) {
+            (*it)->stop();
+        }
+
+        status = status::stopped;
+    }
+
+    void join() {
+        for (auto &collector : collectors) {
+            collector->join();
+        }
+        scheduler.join();
+    }
+
+    private:
+    std::vector<std::unique_ptr<collector>> collectors;
+    std::vector<std::unique_ptr<exporter>> exporters;
+    class recorder recorder;
+    class scheduler scheduler;
+    std::mutex m;
+    enum status status;
+};
+
+inline profiler::profiler()
+    : collectors{},
+      exporters{},
+      recorder{},
+      scheduler(recorder, exporters, std::chrono::nanoseconds{10000000}),
```

Comment:
It's the time it sleeps between its actions, which isn't _quite_ the same as interval, but yeah, that's its point. This will be configuration eventually, like `ddtrace.profiling.scheduler.interval` or such.